playbooks/byo/openshift-master/config.yml

```diff
@@ -4,6 +4,31 @@
 
 - name: Configure master instances
   hosts: masters
+  pre_tasks:
+  - command: hostname -f
+    register: hostname_result
+    when: openshift_hostname is not defined
+
+  - set_fact: openshift_hostname={{ hostname_result.stdout }}
+    when: openshift_hostname is not defined and hostname_result['stdout'] is defined
+
+  - set_fact: openshift_public_hostname={{ openshift_hostname }}
+    when: openshift_public_hostname is not defined
+
+  - set_fact: openshift_ip={{ ansible_default_ipv4.address }}
+    when: openshift_ip is not defined
+
+  - set_fact: openshift_public_ip={{ openshift_ip }}
+    when: openshift_public_ip is not defined
+  vars:
+    openshift_etcd_leader: "{% if groups['etcd-leader'] is defined %}http://{{ hostvars
+          | oo_select_keys(groups['etcd-leader'])
+          | oo_collect(attribute='ansible_default_ipv4.address')
+          | first }}:4001{% endif %}"
+    openshift_etcd_peers: "{% if groups['etcd-peers']|length == 3 %}http://{{ hostvars
+          | oo_select_keys(groups['etcd-peers'])
+          | oo_collect(attribute='ansible_default_ipv4.address')
+          | join(\",\") }}:4001{% endif %}"
```

Comment:
The way facts are being gathered and collected is currently leading to issues with variable precedence, causing some of the templates to not be evaluated.
I'm currently working on a facts module that will set most of the defaults instead of relying on group_vars to set them. Once that is in place, then we should be able to avoid the precedence issues we are seeing now (hopefully, at least).